### DISCUSSION TITLE
Drawer close reconstructs entire component tree

### DISCUSSION THREAD

**Discussion by tcpalmer** [rsuite/rsuite#2330] [Q&A]
When a Drawer is closed, it seems to reconstruct my entire component tree from scratch (console messages show unmount and constructor being called again).  This has the nasty side effect of closing Panels that were open and resetting the sort order of a Table back to the default.  If a user makes these changes, it's a poor experience to have to redo them since they would expect state to persist while the Drawer was viewed.

I assume Modal behaves the same since it looks like Panel uses Modal.  In fact, is this true for any React component that attaches/removes a div directly to body instead of in the DOM tree controlled by the app?

**Comment by SevenOutman** [comment]
Hi @tcpalmer, such behavior of Drawer is actually intended as we use Drawers mostly to contain forms, in which scenario it's expected to reset on close.

It's still easy to keep your states from being reset by declaring them outside of the Drawer.


```jsx
function MyDrawer() {

  // This will not be lost when Drawer is closed
  const userState = useState()

  return (
    <Drawer>
      ...
    </Drawer>
  );
}

```

**Reply by tcpalmer** [reply]
Actually I figured out that it wasn't the ENTIRE component tree being reconstructed but a subset.  That led me to check what I was doing with keys ... and found something stupid.  Problem solved and faith in rsuite restored ;)

Thanks!
